app/routes/questions/actions.tsx

```diff
@@ -0,0 +1,116 @@
+import {useState, useEffect, MouseEvent, useCallback} from 'react'
+import type {ActionArgs} from '@remix-run/cloudflare'
+import {Form, useSearchParams} from '@remix-run/react'
+import {redirect, json} from '@remix-run/cloudflare'
+import {likeQuestion} from '~/server-utils/stampy'
+import {DarkLight, Edit, Flag, Followup, Hide, Like, Search} from '~/components/icons-generated'
+
+export enum ActionType {
+  DARKLIGHT = 'darkLight',
+  EDIT = 'edit',
+  FLAG = 'flag',
+  FOLLOWUP = 'followup',
+  HELPFUL = 'helpful',
+  HIDE = 'hide',
+  SEARCH = 'search',
+}
+export type ActionProps = {
+  Icon: JSX.IntrinsicElements['svg']
+  title: string
+  handler?: (pageid: string) => Promise<string>
+}
+type ActionsDict = {
+  [k: string]: ActionProps
+}
+const actions = {
+  darkLight: {
+    Icon: DarkLight,
+    title: 'Dark/Light',
+  },
+  edit: {
+    Icon: Edit,
+    title: 'Edit',
+  },
+  flag: {
+    Icon: Flag,
+    title: 'Flag',
+  },
+  followup: {
+    Icon: Followup,
+    title: 'Followup',
+  },
+  helpful: {
+    Icon: Like,
+    title: 'Helpful',
+    handler: likeQuestion,
+  },
+  hide: {
+    Icon: Hide,
+    title: 'Hide',
+  },
+  search: {
+    Icon: Search,
+    title: 'Search',
+  },
+} as ActionsDict
+
+export const action = async ({request}: ActionArgs) => {
+  const formData = await request.formData()
+  const pageid = formData.get('pageid') as string
+  const actionType = formData.get('action') as ActionType
+
+  const handler = actions[actionType]?.handler
+  if (handler) {
+    const result = await handler(pageid)
+    if (result != 'ok') return json({error: result}, {status: 400})
+  } else {
+    console.log(`Got unhandled action: ${actionType} for page ${pageid}`)
+  }
+
+  const state = formData.get('stateString')
+  if (state) return redirect(`/?state=${state}`)
+  return redirect('/')
+}
+
+export const Action = ({pageid, actionType}: {pageid: string; actionType: ActionType}) => {
+  const [remixSearchParams] = useSearchParams()
+  const [stateString] = useState(() => remixSearchParams.get('state') ?? '')
+  const {Icon, title} = actions[actionType]
+
+  // Get the state of this action for the given `pageid` from localstorage. This
+  // will result in each action only being allowed once per browser
```

Comment:
Or would it be better to just use the basic react state?

Reply:
hm,, localstorage sounds appropriate, let's see how it works and change only if we discover some problematic edge case